src/integrationTest/java/org/opensearch/security/http/JwtAuthenticationTests.java

```diff
@@ -0,0 +1,253 @@
+	private static final String PUBLIC_KEY = new String(Base64.getEncoder().encode(KEY_PAIR.getPublic().getEncoded()));
+
+	static final TestSecurityConfig.User ADMIN_USER = new TestSecurityConfig.User("admin").roles(ALL_ACCESS);
+
+	private static final String JWT_AUTH_HEADER = "jwt-auth";
+
+	private static final JwtAuthorizationHeaderFactory tokenFactory = new JwtAuthorizationHeaderFactory(
+		KEY_PAIR.getPrivate(),
+		CLAIM_USERNAME,
+		CLAIM_ROLES,
+		JWT_AUTH_HEADER);
+
+	public static final TestSecurityConfig.AuthcDomain JWT_AUTH_DOMAIN = new TestSecurityConfig
+		.AuthcDomain("jwt", BASIC_AUTH_DOMAIN_ORDER - 1)
+		.jwtHttpAuthenticator(new JwtConfigBuilder().jwtHeader(JWT_AUTH_HEADER).signingKey(PUBLIC_KEY).subjectKey(CLAIM_USERNAME)
+			.rolesKey(CLAIM_ROLES))
+		.backend("noop");
+	public static final String SONG_ID_1 = "song-id-01";
+
+	public static final Role DEPARTMENT_SONG_LISTENER_ROLE = new Role("department-song-listener-role")
+		.indexPermissions("indices:data/read/search").on(DEPARTMENT_SONG_INDEX_PATTERN);
+
+	@ClassRule
+	public static final LocalCluster cluster = new LocalCluster.Builder()
+		.clusterManager(ClusterManager.SINGLENODE).anonymousAuth(false)
+		.nodeSettings(Map.of("plugins.security.restapi.roles_enabled", List.of("user_" + ADMIN_USER.getName()  +"__" + ALL_ACCESS.getName())))
+		.authc(AUTHC_HTTPBASIC_INTERNAL).users(ADMIN_USER).roles(DEPARTMENT_SONG_LISTENER_ROLE)
+		.authc(JWT_AUTH_DOMAIN)
+		.build();
+
+	@BeforeClass
+	public static void createTestData() {
+		try (Client client = cluster.getInternalNodeClient()) {
+			client.prepareIndex(QA_SONG_INDEX_NAME).setId(SONG_ID_1).setRefreshPolicy(IMMEDIATE).setSource(SONGS[0]).get();
+		}
+		try(TestRestClient client = cluster.getRestClient(ADMIN_USER)){
+			client.createRoleMapping(ROLE_VP, DEPARTMENT_SONG_LISTENER_ROLE.getName());
+		}
+	}
+
+	@Test
+	public void shouldAuthenticateWithJwtToken_positive() {
+		try(TestRestClient client = cluster.getRestClient(tokenFactory.validToken(USER_SUPERHERO))){
+
+			HttpResponse response = client.getAuthInfo();
+
+			response.assertStatusCode(200);
+			String username = response.getTextFromJsonBody(POINTER_USERNAME);
+			assertThat(username, equalTo(username));
+		}
+	}
+
+	@Test
+	public void shouldAuthenticateWithJwtToken_positiveWithAnotherUsername() {
+		try(TestRestClient client = cluster.getRestClient(tokenFactory.validToken(USERNAME_ROOT))){
+
+			HttpResponse response = client.getAuthInfo();
+
+			response.assertStatusCode(200);
+			String username = response.getTextFromJsonBody(POINTER_USERNAME);
+			assertThat(username, equalTo(USERNAME_ROOT));
+		}
+	}
+
+	@Test
+	public void shouldAuthenticateWithJwtToken_failureLackingUserName() {
+		try(TestRestClient client = cluster.getRestClient(tokenFactory.tokenWithoutPreferredUsername(USER_SUPERHERO))){
+
+			HttpResponse response = client.getAuthInfo();
+
+			response.assertStatusCode(401);
```

Comment:
Many of these tests are verifying 401 responses, is there any way to better verify the failure condition, e.g. check a log for an entry that stays 'invalid token, did not contain a user name'?

Reply:
Done, the only possibility was in fact to check logs.